govcd/security_tags.go

```diff
@@ -0,0 +1,154 @@
+	if err != nil {
+		return nil, err
+	}
+
+	securityTaggedEntities := make([]*types.SecurityTaggedEntity, len(rawValues))
+
+	for i, k := range rawValues {
+		var temp types.SecurityTaggedEntity
+		err = json.Unmarshal(k, &temp)
+		if err != nil {
+			return nil, fmt.Errorf("error when unmarshalling SecurityTaggedEntity - %s", err)
+		}
+		securityTaggedEntities[i] = &temp
+	}
+	return securityTaggedEntities, nil
+}
+
+// Retrieves the list of security tags that are in the organization and can be reused to tag an entity.
+// The list of tags include tags assigned to entities within the organization.
+//This API is meant for organization user only (i.e. not system provider).
+func (org *Org) GetSecurityTagValues(filter string) ([]*types.SecurityTagValue, error) {
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointSecurityTags
+	apiVersion, err := org.client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	urlRef, err := org.client.OpenApiBuildEndpoint(endpoint, "/values")
+	if err != nil {
+		return nil, err
+	}
+
+	// If sysadmin, getting the org context and this method only works for org users
+	orgContextHeaders := make(map[string]string)
+	if org.client.IsSysAdmin {
+		orgContextHeaders["X-VMWARE-VCLOUD-AUTH-CONTEXT"] = org.Org.Name
+		orgContextHeaders["X-VMWARE-VCLOUD-TENANT-CONTEXT"] = org.Org.ID
+	}
+
+	v := url.Values{}
+	v.Set("filter", filter)
+	rawValues, err := org.client.openApiGetAllPages(apiVersion, urlRef, v, nil, nil, orgContextHeaders)
+	if err != nil {
+		return nil, err
+	}
+
+	securityTaggedValues := make([]*types.SecurityTagValue, len(rawValues))
+
+	for i, k := range rawValues {
+		var temp types.SecurityTagValue
+		err = json.Unmarshal(k, &temp)
+		if err != nil {
+			return nil, fmt.Errorf("error when unmarshalling SecurityTaggedEntity - %s", err)
+		}
+		securityTaggedValues[i] = &temp
+	}
+
+	return securityTaggedValues, nil
+}
+
+// GetVMTags Retrieves the list of tags for a specific VM. If user has view right to the VM, user can view its tags.
+func (org *Org) GetVMSecurityTags(id string) (*types.EntitySecurityTags, error) {
+	endpoint := types.OpenApiPathVersion1_0_0 + types.OpenApiEndpointSecurityTags
+	apiVersion, err := org.client.getOpenApiHighestElevatedVersion(endpoint)
+	if err != nil {
+		return nil, err
+	}
+
+	urlRef, err := org.client.OpenApiBuildEndpoint(endpoint, fmt.Sprintf("/vm/%s", id))
+	if err != nil {
+		return nil, err
+	}
+
+	var entitySecurityTags types.EntitySecurityTags
+	err = org.client.OpenApiGetItem(apiVersion, urlRef, nil, &entitySecurityTags, nil)
+	if err != nil {
+		return nil, err
+	}
+
+	return &entitySecurityTags, nil
+}
+
+// UpdateSecurityTag updates the entities associated with a Security Tag.
+// Only the list of tagged entities can be updated. The name cannot be updated.
+// Any other existing entities not in the list will be untagged.
+func (org *Org) UpdateSecurityTag(securityTag *types.SecurityTag) error {
```

Comment:
The API itself doesn't return anything. The methods that returns something it's because the API does.
Please check https://developer.vmware.com/apis/vmware-cloud-director/v36.3/security-tags/ 
I tried to be as much transparent with the API as possible.

Reply:
We usually take "refresh path". Fetch value after update and return. So it's nice as I don't need to do that manually. You will see that in your test directly.

Reply:
I think I understood what you said, but let me give you more insight about that function/endpoint.
Whatever you set on parameter `securityTag` is the security tag with all VMs associated. You're not just updating one but all. Whatever you place in that parameter, as long as it doesn't error, is what is going to be set. 
No tasks whatsoever involved neither.